moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/tesseract/basic_types.h

```diff
@@ -0,0 +1,285 @@
+  struct PairHash
+  {
+    std::size_t operator()(const LinkNamesPair& pair) const
+    {
+      return std::hash<std::string>()(pair.first + pair.second);
+    }
+  };
+  typedef std::unordered_map<LinkNamesPair, std::string, PairHash> AllowedCollisionEntries;
+  AllowedCollisionEntries lookup_table_;
+
+  /**
+   * @brief Create a pair of strings, where the pair.first is always <= pair.second
+   * @param link_name1 First link name
+   * @param link_name2 Second link anme
+   * @return LinkNamesPair a lexicographically sorted pair of strings
+   */
+  static inline LinkNamesPair makeOrderedLinkPair(const std::string& link_name1, const std::string& link_name2)
+  {
+    if (link_name1 <= link_name2)
+      return std::make_pair(link_name1, link_name2);
+    else
+      return std::make_pair(link_name2, link_name1);
+  }
+
+public:
+  /**
+   * @brief Clears the list of allowed collisions
+   * @return AllowedCollisionEntries an unordered map containing all allowed
+   *         collision entries. The keys of the unordered map are a std::pair
+   *         of the link names in the allowed collision pair.
+   */
+  const AllowedCollisionEntries& getAllAllowedCollisions() const
+  {
+    return lookup_table_;
+  }
+};
+typedef std::shared_ptr<AllowedCollisionMatrix> AllowedCollisionMatrixPtr;
+typedef std::shared_ptr<const AllowedCollisionMatrix> AllowedCollisionMatrixConstPtr;
+
+/**
+ * @brief Should return true if contact allowed, otherwise false.
+ *
+ * Also the order of strings should not matter, the function should handled by the function.
+ */
+typedef std::function<bool(const std::string&, const std::string&)> IsContactAllowedFn;
+
+namespace CollisionObjectTypes
+{
+enum CollisionObjectType
+{
+  UseShapeType = 0, /**< @brief Infer the type from the type specified in the shapes::Shape class */
+
+  // These convert the meshes to custom collision objects
+  ConvexHull =
+      1, /**< @brief Use the mesh in shapes::Shape but make it a convex hulls collision object. (if not convex it will
+            be converted) */
+  MultiSphere = 2, /**< @brief Use the mesh and represent it by multiple spheres collision object */
+  SDF = 3          /**< @brief Use the mesh and rpresent it by a signed distance fields collision object */
+};
+}
+typedef CollisionObjectTypes::CollisionObjectType CollisionObjectType;
+typedef std::vector<CollisionObjectType> CollisionObjectTypeVector;
+
+namespace BodyTypes
+{
+enum BodyType
+{
+  ROBOT_LINK = 0,    /**< @brief These are links at the creation of the environment */
+  ROBOT_ATTACHED = 1 /**< @brief These are links that are added after initial creation */
+};
+}
+typedef BodyTypes::BodyType BodyType;
+
+namespace ContinouseCollisionTypes
+{
+enum ContinouseCollisionType
+{
+  CCType_None,
+  CCType_Time0,
+  CCType_Time1,
+  CCType_Between
+};
+}
+typedef ContinouseCollisionTypes::ContinouseCollisionType ContinouseCollisionType;
+
+namespace ContactTestTypes
```

Comment:
Most of these singe enum namespaces don't need to exist, and should be using `enum class` instead. See [this question](https://stackoverflow.com/a/18335862).